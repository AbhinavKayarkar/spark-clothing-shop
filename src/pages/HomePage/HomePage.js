import React from 'react';
import { Helmet } from 'react-helmet';
import { NavLink } from 'react-router-dom';
import './HomePage.scss';
// functional component using arrow function ( () => {return ();})
const HomePage = () => {
  return (
    <>
      {/* Title of this page (Home) */}
      <Helmet>
        <title>Home</title>
      </Helmet>
      {/* Carousel for homepage */}
      <div id="carouselExampleAutoplaying" className="carousel slide" data-bs-ride="carousel">
        <div className="carousel-inner">
          {/* first item with image and test */}
          <div className="carousel-item active">
            <img
              src="/assets/images/homepage-images/bg-image.jpg"
              className="bd-placeholder-img "
              width="100%"
              height="100%"
              alt="bg-image"
            />
            {/* <div className="card-img-overlay d-flex flex-column justify-content-around">
              <div className="container-sm">
                <h5 className="card-title fs-1 mb-0">NEW SEASON ARRIVALS</h5>
                <p className="card-text fw-bolder fs-5 p-3">CHECK OUT ALL !!!</p>
              </div>
            </div> */}
            <div className="card-img-overlay d-flex flex-column justify-content-center">
              <div className="container-fluid">
                <h3 className="text-break text-black">NEW SEASON ARRIVALS</h3>
                <p className="card-text fw-bolder fs-5 p-3 text-white">CHECK OUT ALL !!!</p>
              </div>
            </div>
          </div>
          {/* second item with image and text */}
          <div className="carousel-item">
            <img
              src="/assets/images/homepage-images/black-friday-sale-img.jpg"
              className="bd-placeholder-img"
              width="100%"
              height="100%"
              alt="black-friday-sale-img"
            />
            <div className="card-img-overlay d-flex flex-column justify-content-end mb-3">
              <div className="container">
                <span
                  className=" btn btn-danger spinner-grow spinner-grow-sm"
                  role="status"
                  aria-hidden="true"></span>
                <span className="card-text fw-bolder fs-5 p-3 text-white">
                  Live Now !!! Hurry Up
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>

      <div className="text-center mt-3">
        <NavLink to="/products" className="btn btn-dark">
          Shop Now
        </NavLink>
      </div>
    </>
  );
};

export default HomePage;
