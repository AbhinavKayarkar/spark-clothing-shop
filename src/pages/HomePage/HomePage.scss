/* Only for home page carousel component */
.carousel .carousel-item {
  height: 550px;
}

/* carousel img css starts here */
.carousel-item img {
  position: absolute;
  object-fit: fill;
  top: 0;
  left: 0;
  min-height: 100%;
}

/* carousel img css ends here */