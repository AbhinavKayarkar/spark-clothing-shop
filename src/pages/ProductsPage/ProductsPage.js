import React, { useEffect, useState } from 'react'; // Hooks for handling the fn's
import { NavLink, Link, useLocation, Outlet } from 'react-router-dom'; // navigation or routing
import { Helmet } from 'react-helmet'; // providing title for pages
import axios from 'axios'; // client tool for backend
import TabProducts from './TabProducts/TabProducts';
import { AiTwotoneStar } from 'react-icons/ai'; // icons for designing
import './ProductsPage.scss';

const useQuery = () => new URLSearchParams(useLocation().search);

const ProductsPage = () => {
  const [isLoading, setIsLoading] = useState(true); // store data belongs to loading status
  const [isError, setIsError] = useState(false); // store data belongs to error status
  const [productList, setProductList] = useState([]); // store data belongs to productList status
  let api = null;
  let query = useQuery();
  let queryParam = query.get('category');

  if (queryParam === 'men') {
    api = 'http://localhost:8000/products?category=men';
  } else if (queryParam === 'women') {
    api = 'http://localhost:8000/products?category=women';
  } else if (queryParam === 'kid') {
    api = 'http://localhost:8000/products?category=kid';
  } else {
    api = 'http://localhost:8000/products';
  }

  useEffect(() => {
    axios //used to connect with backend .get is retrive the data from backend.
      .get(api) // data is stored in db.json
      .then((res) => {
        //console.log(res.data);
        setIsLoading(false);
        setIsError(false);
        setProductList(res.data);
      })
      .catch((err) => {
        // catch is used to handle the errors.
        //console.log(err);
        setIsLoading(false);
        setIsError(true);
        setProductList([]);
      })
      .finally(() => {
        // finally is always executed at the end.
        //console.log(' It is Over!');
      });
  }, [api]);

  // spinner
  if (isLoading) {
    // if isLoading is true then this code snippet is excuted and loading sign.
    return (
      <div className="text-center">
        {/* spinner with center - bootstrap*/}
        <div className="spinner-border text-danger" role="status"></div>
      </div>
    );
  }

  // Error Message
  if (isError) {
    // if any Error is Occur then this block is executed.
    return (
      <div className="alert alert-danger" role="alert">
        Sorry! Unable to fetch. Try again later.
      </div>
    );
  }

  // If error & loading is false, then this return block will be render on app
  return (
    <>
      {/* Title of this page (Products) */}
      <Helmet>
        <title>Products</title>
      </Helmet>

      {/* Heading Text */}
      <h1 className="text-center text-info">Our Collection</h1>

      <div className="container-fluid mx-2">
        <div className="row mt-5 mx-2">
          {/* Categorywise Displaying Product */}
          <div className="col-md-3">
            <TabProducts />
          </div>

          {/* All Products Display */}
          <div className="col-md-9">
            <div className="row">
              {productList?.map((product) => {
                return (
                  <div className="col-md-4" key={product.id}>
                    <div className="card">
                      <img
                        src={product.imgSrc}
                        className="card-img-top"
                        width={280}
                        height={450}
                        alt={product.title}
                      />
                      <p className="product-Label">{product.label}</p>
                      <div className="product-review">
                        <p className="rating">
                          {product.rating}
                          {<AiTwotoneStar />}
                          {product.review}
                        </p>
                      </div>
                      <div className="card-body">
                        <h5 className="card-title">
                          <Link to={`/products/${product.id}`} className="product-details">
                            {product.title}
                          </Link>
                        </h5>
                        <p className="card-text text-secondary">{product.description}</p>
                        <span className="me-3 fs-5">{product.sellingPrice}</span>
                        <span className="me-2 text-secondary text-decoration-line-through">
                          {product.maxretailPrice}
                        </span>
                        <span className="orange-text">{product.discount}</span>
                      </div>
                    </div>
                  </div>
                );
              })}
            </div>
          </div>
          <Outlet />
        </div>
      </div>
    </>
  );
};

export default ProductsPage;
