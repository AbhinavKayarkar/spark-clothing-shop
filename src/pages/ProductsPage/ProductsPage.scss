// Products Page css -- start
.orange-text {
  color: lightsalmon;
}

.product-review {
  position: absolute;
  bottom: 150px;
  left: 30px;
  background-color: aliceblue;
}

.rating {
  padding: 0 4px;
  margin-bottom: 0px;
  text-align: center;
}

.product-Label {
  position: absolute;
  top: 20px;
  left: 20px;
  background-color: rgb(239, 43, 76);
  padding: 0 4px;
  margin-bottom: 0px;
  text-align: center;
}

.product-details {
  color: black;
  text-decoration: none;
}

// Products page css ends here