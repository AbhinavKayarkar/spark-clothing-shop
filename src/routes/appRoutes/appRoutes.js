import React, { Suspense } from 'react';
import { Route, Routes } from 'react-router-dom';

// component imports
import HomePage from '../../pages/HomePage/HomePage';
import PageNotFound from '../../components/PageNotFound/PageNotFound';
import ProductsPage from '../../pages/ProductsPage/ProductsPage';
import ProductDetails from '../../pages/ProductsPage/ProductDetails/ProductDetails';
import Feedback from '../../pages/ProductsPage/Feedback/Feedback';
import AboutUsContent from '../../pages/AboutUsPage/AboutUsContent/AboutUsContent';

// Lazy Loading
const AboutUs = React.lazy(() => import('../../pages/AboutUsPage/AboutUs'));
const ContactUsPage = React.lazy(() => import('../../pages/ContactUsPage/ContactUsPage'));
const ContactUsContent = React.lazy(() =>
  import('../../pages/ContactUsPage/ContactUsContent/ContactUsContent')
);

export const appRoutes = (
  <Suspense fallback={<div className="spinner-border text-warning"></div>}>
    <Routes>
      <Route path="*" element={<PageNotFound />}></Route>
      <Route path="/" exact element={<HomePage />}></Route>
      <Route path="/products" element={<ProductsPage />}></Route>
      <Route path="products/:id" element={<ProductDetails />}></Route>
      <Route path="products/:id/feedback" element={<Feedback />}></Route>
      <Route path="/about-us/" element={<AboutUs />}>
        <Route path="content" element={<AboutUsContent />}></Route>
      </Route>
      <Route path="/contact-us" element={<ContactUsPage />}></Route>
      <Route path="/contact-us/content" element={<ContactUsContent />}></Route>
    </Routes>
  </Suspense>
);
