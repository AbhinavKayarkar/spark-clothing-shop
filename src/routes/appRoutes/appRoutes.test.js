import { cleanup, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import HomePage from '../../pages/HomePage/HomePage';
import ProductsPage from '../../pages/ProductsPage/ProductsPage';
import { appRoutes } from './appRoutes';

import renderer from 'react-test-renderer';

// mocking the components
jest.mock('../../pages/HomePage/HomePage');
jest.mock('../../pages/ProductsPage/ProductsPage');

afterEach(cleanup);

// test suites
describe('appRoutes', () => {
  test('should the homepage component render properly', () => {
    // Arrange
    HomePage.mockImplementation(() => <div>HomePageMock</div>);

    // Act
    render(
      <MemoryRouter>
        <HomePage />
      </MemoryRouter>
    );

    // Assert
    expect(screen.getByText('HomePageMock')).toBeInTheDocument();
  });

  test('should the ProductsPage component render properly', () => {
    // Arrange
    ProductsPage.mockImplementation(() => <div>ProductsPage</div>);

    // Act
    render(
      <MemoryRouter>
        <ProductsPage />
      </MemoryRouter>
    );

    // Assert
    expect(screen.getByText('ProductsPage')).toBeInTheDocument();
  });

  // Lazy Loading
  // testing about page
  it('renders the AboutUs Page component correctly', async () => {
    render(<MemoryRouter initialEntries={['/about-us']}>{appRoutes}</MemoryRouter>);

    const aboutUs = await screen.findByText('The Tagline');
    expect(aboutUs).toBeInTheDocument();
  });

  // testing ContactUs page
  it('renders the ContactUs component correctly', async () => {
    render(<MemoryRouter initialEntries={['/contact-us']}>{appRoutes}</MemoryRouter>);

    const contactUsPage = await screen.findByText('Contact Us');
    expect(contactUsPage).toBeInTheDocument();
  });

  // testing ContactUsContent page
  it('renders the ContactUsContent component correctly', async () => {
    render(<MemoryRouter initialEntries={['/contact-us/content']}>{appRoutes}</MemoryRouter>);

    const contactUsContent = await screen.findByText('Get In Touch Form');
    expect(contactUsContent).toBeInTheDocument();
  });

  // testing snapshot to take whole approutes code as copy print.
  it('should render correctly', () => {
    const snapshotInJson = renderer.create(<appRoutes />).toJSON();
    expect(snapshotInJson).toMatchSnapshot();
  });
});
