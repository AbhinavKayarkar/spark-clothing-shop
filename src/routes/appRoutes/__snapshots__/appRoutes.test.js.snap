// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appRoutes should render correctly 1`] = `<appRoutes />`;
